common/lib/plugins/efm2/monitor.ts

```diff
@@ -0,0 +1,310 @@
+            const monitorContext = monitorContextRef.deref();
+            if (monitorContext == null) {
+              continue;
+            }
+
+            if (this.hostUnhealthy) {
+              // Kill connection
+              monitorContext.setHostUnhealthy(true);
+              monitorContext.setInactive();
+              const connectionToAbort = monitorContext.getClient();
+              if (connectionToAbort != null) {
+                await this.endMonitoringClient();
+                this.abortedConnectionsCounter.inc();
+              }
+            } else if (monitorContext.isActive()) {
+              tmpActiveContexts.push(monitorContextRef);
+            }
+
+            // activeContexts is empty now and tmpActiveContexts contains all yet active contexts
+            // Add active contexts back to the queue.
+            this.activeContexts.push(...tmpActiveContexts);
+
+            const delayNanos = this.failureDetectionIntervalNanos - (statusCheckEndTimeNanos - statusCheckStartTimeNanos);
+            await sleep(Math.max(delayNanos, MonitorImpl.TASK_SLEEP_NANOS) / 1000000);
+          }
+        } catch (error: any) {
+          logger.debug(Messages.get("MonitorImpl.exceptionDuringMonitoringContinue", error.message));
+        }
+      }
+    } catch (error: any) {
+      logger.debug(Messages.get("MonitorImpl.exceptionDuringMonitoringStop", error.message));
+    } finally {
+      await this.endMonitoringClient();
+      await sleep(3000);
+    }
+
+    logger.debug(Messages.get("MonitorImpl.stopMonitoring", this.hostInfo.host));
+  }
+
+  /**
+   * Check the status of the monitored server by sending a ping.
+   *
+   * @return whether the server is still alive and the elapsed time spent checking.
+   */
+  async checkConnectionStatus(): Promise<boolean> {
+    const connectContext = this.telemetryFactory.openTelemetryContext("Connection status check", TelemetryTraceLevel.FORCE_TOP_LEVEL);
+    connectContext.setAttribute("url", this.hostInfo.host);
+    try {
+      if (!(await this.pluginService.isClientValid(this.monitoringClient))) {
+        // Open a new connection.
+        const monitoringConnProperties: Map<string, any> = new Map(this.properties);
+
+        for (const key of this.properties.keys()) {
+          if (!key.startsWith(WrapperProperties.MONITORING_PROPERTY_PREFIX)) {
+            continue;
+          }
+
+          monitoringConnProperties.set(key.substring(WrapperProperties.MONITORING_PROPERTY_PREFIX.length), this.properties.get(key));
+          monitoringConnProperties.delete(key);
+        }
+
+        logger.debug(`Opening a monitoring connection to ${this.hostInfo.url}`);
+        this.monitoringClient = await this.pluginService.forceConnect(this.hostInfo, monitoringConnProperties);
+        logger.debug(`Successfully opened monitoring connection to ${this.monitoringClient.id} - ${this.hostInfo.url}`);
+        return true;
+      }
+      return true;
+    } catch (error: any) {
```

Comment:
Do we need to try closing the monitoring client?

Reply:
we only want to close the monitoring client when monitoring is stopped. it may continue to monitor after this check